data/transactions/logic/assembler.go

```diff
@@ -1182,13 +1223,359 @@ const AssemblerDefaultVersion = 1
+	// since we're using a stable sort, constants with the same frequency
+	// will retain their current ordering (i.e. first referenced, first in int block)
+	sort.SliceStable(freqs, func(i, j int) bool {
+		return freqs[i].freq > freqs[j].freq
+	})
+
+	// sort refs from last to first in raw
+	// this way when we iterate through them and potentially change the size of raw,
+	// the later positions will not affect the indexes of the earlier positions
+	sort.Slice(ops.intcRefs, func(i, j int) bool {
+		return ops.intcRefs[i].position > ops.intcRefs[j].position
+	})
+
+	opIntc0 := OpsByName[ops.Version]["intc_0"].Opcode
+	opIntc1 := OpsByName[ops.Version]["intc_1"].Opcode
+	opIntc2 := OpsByName[ops.Version]["intc_2"].Opcode
+	opIntc3 := OpsByName[ops.Version]["intc_3"].Opcode
+	opIntc := OpsByName[ops.Version]["intc"].Opcode
+	opPushInt := OpsByName[ops.Version]["pushint"].Opcode
+
+	var scratch [binary.MaxVarintLen64]byte
+	raw := ops.pending.Bytes()
+	for _, ref := range ops.intcRefs {
+		singleton := false
+		newIndex := -1
+		for i, f := range freqs {
+			if ref.value == f.value {
+				singleton = f.freq == 1
+				newIndex = i
+				break
+			}
+		}
+		if newIndex == -1 {
+			return ops.errorf("Value not found in intc block: %d", ref.value)
+		}
+
+		var newBytes []byte
+		if singleton {
+			vlen := binary.PutUvarint(scratch[:], ref.value)
+			newBytes = make([]byte, 1+vlen)
+			newBytes[0] = opPushInt
+			copy(newBytes[1:], scratch[:vlen])
+		} else {
+			switch newIndex {
+			case 0:
+				newBytes = []byte{opIntc0}
+			case 1:
+				newBytes = []byte{opIntc1}
+			case 2:
+				newBytes = []byte{opIntc2}
+			case 3:
+				newBytes = []byte{opIntc3}
+			default:
+				newBytes = []byte{opIntc, uint8(newIndex)}
+			}
+		}
+
+		var currentBytesLen int
+		switch raw[ref.position] {
+		case opIntc0, opIntc1, opIntc2, opIntc3:
+			currentBytesLen = 1
+		case opIntc:
+			currentBytesLen = 2
+		default:
+			return ops.errorf("Unexpected op at intReference: %d, from source line %d", raw[ref.position], ops.OffsetToLine[ref.position])
+		}
+
+		raw = replaceBytes(raw, ref.position, currentBytesLen, newBytes)
+		positionDelta := len(newBytes) - currentBytesLen
+
+		for i := range ops.bytecRefs {
+			if ops.bytecRefs[i].position > ref.position {
+				ops.bytecRefs[i].position += positionDelta
+			}
+		}
+
+		for label := range ops.labels {
+			if ops.labels[label] > ref.position {
+				ops.labels[label] += positionDelta
+			}
+		}
+
+		for i := range ops.labelReferences {
+			if ops.labelReferences[i].position > ref.position {
+				ops.labelReferences[i].position += positionDelta
+			}
+		}
+
+		fixedOffsetsToLine := make(map[int]int, len(ops.OffsetToLine))
+		for pos, sourceLine := range ops.OffsetToLine {
+			if pos > ref.position {
+				fixedOffsetsToLine[pos+positionDelta] = sourceLine
+			} else {
+				fixedOffsetsToLine[pos] = sourceLine
+			}
+		}
+		ops.OffsetToLine = fixedOffsetsToLine
+	}
+
+	ops.pending.Reset()
+	ops.pending.Write(raw)
```

Comment:
tiny nit: I think it may be slightly more performant to do `ops.pending = bytes.NewBuffer(raw)` versus `Write()` which will trigger some slice-growing stuff based on `len(raw)`

Reply:
unless someone else is holding on to a reference to `pending` ... I don't know much about the code or how it's used

Reply:
can you tell I used to be a C++ programmer who was always looking for opportunities to call `.swap()` on data structures